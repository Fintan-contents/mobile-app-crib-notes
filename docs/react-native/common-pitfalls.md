---
title: ハマりがちな落とし穴
---

React Nativeで開発する際に起きた困り事や不具合についてまとめています。
