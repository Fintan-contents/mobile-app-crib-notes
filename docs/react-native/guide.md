---
title: 何らかのガイドを書きたい気持ちはある
---
