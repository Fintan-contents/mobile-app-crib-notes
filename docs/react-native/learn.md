---
title: About
hide_table_of_contents: true
---

このコンテンツはReact Nativeでのモバイルアプリケーション開発を学習するためのコンテンツです。

ReactでWebアプリケーションを実装できるがReact Nativeでモバイルアプリケーションを実装したことがないアプリケーション開発者を対象として、次のことについて学ぶことを目的としています。

- ReactとReact Nativeの違い
- React Nativeで作成したモバイルアプリケーションの動作確認方法

また、作成したモバイルアプリケーションの配布方法やテスト手法についてはこのコンテンツでは触れません。

<!-- textlint-disable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->

import {PageList} from '../../src/components';
const overviews = [
  {
    title: 'Getting Started',
    to: '/react-native/learn/getting-started',
    summary: <>React Nativeでのアプリ開発に必要な環境を用意します。</>
  },
  {
    title: 'アプリの実装前に',
    to: '/react-native/learn/basic-concepts',
    summary: <>React Nativeでアプリを実装し始める前に最低限必要となる前提知識や、React NativeとReact Navigationの基本的な知識について学びます。</>
  },
  {
    title: 'ToDoアプリの実装',
    to: '/react-native/learn/todo-app',
    summary: <>React Nativeで簡単なToDoアプリを実装します。</>
  },
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->
