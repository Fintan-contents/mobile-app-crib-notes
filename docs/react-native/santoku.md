---
title: Example Application
sidebar_label: About
hide_table_of_contents: true
---

React Nativeサンプルアプリケーションの設計ドキュメントです。

<!-- textlint-disable ja-technical-writing/sentence-length -->

import {PageList} from '../../src/components';
const overviews = [
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length -->
