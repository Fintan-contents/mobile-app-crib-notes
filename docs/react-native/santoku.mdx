---
title: Example Application
sidebar_label: About
hide_table_of_contents: true
---

<!-- textlint-disable ja-technical-writing/sentence-length -->

import {PageList} from '../../src/components';
const overviews = [
  {
    title: 'Development',
    summary: '開発環境やビルドに関するドキュメント',
    to: '/react-native/santoku/development',
  }
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length -->
