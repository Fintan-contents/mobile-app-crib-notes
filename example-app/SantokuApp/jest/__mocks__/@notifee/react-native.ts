import notifee from '@notifee/react-native';
export * from '@notifee/react-native/dist/types/Library';
export * from '@notifee/react-native/dist/types/Notification';
export * from '@notifee/react-native/dist/types/Trigger';
export * from '@notifee/react-native/dist/types/NotificationIOS';
export * from '@notifee/react-native/dist/types/NotificationAndroid';

const mock: typeof notifee = {
  cancelAllNotifications: jest.fn(),
  cancelDisplayedNotifications: jest.fn(),
  cancelTriggerNotifications: jest.fn(),
  cancelNotification: jest.fn(),
  cancelDisplayedNotification: jest.fn(),
  cancelTriggerNotification: jest.fn(),
  createChannel: jest.fn(),
  createChannels: jest.fn(),
  createChannelGroup: jest.fn(),
  createChannelGroups: jest.fn(),
  deleteChannel: jest.fn(),
  deleteChannelGroup: jest.fn(),
  displayNotification: jest.fn(),
  openAlarmPermissionSettings: jest.fn(),
  createTriggerNotification: jest.fn(),
  getTriggerNotificationIds: jest.fn(),
  getDisplayedNotifications: jest.fn(),
  getTriggerNotifications: jest.fn(),
  getChannel: jest.fn(),
  isChannelCreated: jest.fn(),
  isChannelBlocked: jest.fn(),
  getChannels: jest.fn(),
  getChannelGroup: jest.fn(),
  getChannelGroups: jest.fn(),
  getInitialNotification: jest.fn(),
  onBackgroundEvent: jest.fn(),
  onForegroundEvent: jest.fn(),
  openNotificationSettings: jest.fn(),
  registerForegroundService: jest.fn(),
  stopForegroundService: jest.fn(),
  requestPermission: jest.fn(),
  setNotificationCategories: jest.fn(),
  getNotificationCategories: jest.fn(),
  getNotificationSettings: jest.fn(),
  getBadgeCount: jest.fn(),
  setBadgeCount: jest.fn(),
  incrementBadgeCount: jest.fn(),
  decrementBadgeCount: jest.fn(),
  openBatteryOptimizationSettings: jest.fn(),
  isBatteryOptimizationEnabled: jest.fn(),
  getPowerManagerInfo: jest.fn(),
  openPowerManagerSettings: jest.fn(),
  hideNotificationDrawer: jest.fn(),
  SDK_VERSION: '5.4.1',
};

// 複数のファイルでnotifeeをimportしていた場合に、redefineされないようにモックが存在するか確認
if (!__mocks.notifee) {
  Object.defineProperty(__mocks, 'notifee', {value: mock});
}

// テストケースごとにモックは初期化しておく。
beforeEach(() =>
  Object.values(mock).forEach(fn => {
    if (jest.isMockFunction(fn)) {
      fn.mockClear();
    }
  }),
);

export default mock;
