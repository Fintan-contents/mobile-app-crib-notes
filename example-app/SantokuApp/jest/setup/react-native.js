// https://reactnavigation.org/docs/testing#mocking-native-modules
// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

// !!! NOTICE !!!
// The React Navigation documentation says that `react-native-reanimated` and `react-native-gesture-handler`
// also need to be configured with mocks.
// But since React Navigation v6 does not already depend on those packages, mocking those packages is not necessary
// for React Navigation v6.
//
// If you want to use `react-native-reanimated` and `react-native-gesture-handler`, add mock according to the guide
// provided by each package.
// - `react-native-reanimated`: https://docs.swmansion.com/react-native-reanimated/docs/guide/testing/
// - `react-native-gesture-handler`: https://docs.swmansion.com/react-native-gesture-handler/docs/guides/testing
