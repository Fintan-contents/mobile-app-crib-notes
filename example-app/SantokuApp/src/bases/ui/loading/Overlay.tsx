import React, {useCallback, useEffect, useRef, useState} from 'react';
import {Animated, Dimensions, StyleProp, StyleSheet, ViewStyle} from 'react-native';

import EndCallback = Animated.EndCallback;

export type OverlayProps = {
  visible: boolean;
  onHideEnd?: () => void;
  fadeDuration?: number;
  style?: StyleProp<ViewStyle>;
};

export const Overlay: React.FC<React.PropsWithChildren<OverlayProps>> = ({
  visible,
  onHideEnd,
  fadeDuration,
  style,
  ...props
}) => {
  const fadeAnim = useRef(new Animated.Value(0)).current;
  const [render, setRender] = useState(false);

  const fade = useCallback(
    (toValue: number, callback?: EndCallback) => {
      const animation = Animated.timing(fadeAnim, {
        toValue,
        duration: fadeDuration,
        useNativeDriver: true,
      });
      animation.start(result => {
        callback?.(result);
      });
    },
    [fadeAnim, fadeDuration],
  );

  const show = useCallback(() => {
    fade(1);
    setRender(true);
  }, [fade]);

  const hide = useCallback(() => {
    fade(0, () => {
      setRender(false);
      if (onHideEnd) {
        onHideEnd();
      }
    });
  }, [fade, onHideEnd]);

  useEffect(() => {
    if (visible) {
      if (!render) {
        show();
      }
      return;
    }
    if (render) {
      hide();
    }
  }, [visible, render, show, hide]);

  const overlayStyle = StyleSheet.flatten([{opacity: fadeAnim}, styles.overlay, style]);

  return (
    <>
      {render && (
        <Animated.View style={overlayStyle} testID="overlayAnimatedView">
          {props.children}
        </Animated.View>
      )}
    </>
  );
};

Overlay.defaultProps = {
  fadeDuration: 200,
};

const styles = StyleSheet.create({
  overlay: {
    flex: 1,
    position: 'absolute',
    bottom: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.2)',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: Dimensions.get('window').height,
  },
});
