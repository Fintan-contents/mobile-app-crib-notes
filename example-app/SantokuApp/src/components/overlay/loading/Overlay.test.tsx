import {act} from '@testing-library/react-hooks';
import {render} from '@testing-library/react-native';
import React from 'react';
import {Text, ViewStyle} from 'react-native';
import {ReactTestInstance} from 'react-test-renderer';

import {Overlay} from './Overlay';

jest.useFakeTimers();

function getStyle<T>(instance: ReactTestInstance) {
  return instance.props.style as T;
}

const ChildComponent = () => {
  return <Text testID="text">test</Text>;
};

const FADE_DURATION = 200;

describe('Overlay', () => {
  it('Overlayが正常にrenderできることを確認', () => {
    const renderResult = render(
      <Overlay visible>
        <ChildComponent />
      </Overlay>,
    );

    const {queryByTestId, queryByText, getByTestId} = renderResult;

    expect(queryByTestId('text')).not.toBeNull();
    expect(queryByText('test')).not.toBeNull();

    expect(getStyle<ViewStyle>(getByTestId('overlayAnimatedView')).opacity).toBe(0);
    expect(renderResult).toMatchSnapshot('render直後');

    act(() => {
      jest.advanceTimersByTime(FADE_DURATION);
    });

    expect(getStyle<ViewStyle>(getByTestId('overlayAnimatedView')).opacity).toBe(1);
    expect(renderResult).toMatchSnapshot('フェードイン後');

    renderResult.rerender(
      <Overlay visible={false}>
        <ChildComponent />
      </Overlay>,
    );

    act(() => {
      jest.advanceTimersByTime(FADE_DURATION);
    });

    expect(queryByTestId('overlayAnimatedView')).toBeNull();
    expect(renderResult).toMatchSnapshot('フェードアウト後');
  });

  it('Overlayに指定したpropsがrenderに反映されていることを確認', () => {
    const onHideEnd = jest.fn();
    // fadeDurationはテストコードでアニメーション時間を調整できないため検証対象外
    const {queryByText, getByTestId} = render(
      <Overlay visible onHideEnd={onHideEnd} style={{backgroundColor: 'red'}}>
        <ChildComponent />
      </Overlay>,
    );

    act(() => {
      jest.advanceTimersByTime(FADE_DURATION);
    });

    expect(queryByText('test')).not.toBeNull();
    expect(getStyle<ViewStyle>(getByTestId('overlayAnimatedView')).backgroundColor).toBe('red');
  });
});
