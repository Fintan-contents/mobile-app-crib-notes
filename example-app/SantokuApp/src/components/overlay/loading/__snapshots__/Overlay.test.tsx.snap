// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay Overlayが正常にrenderできることを確認: render直後 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(0, 0, 0, 0.2)",
      "bottom": 0,
      "flex": 1,
      "height": 1334,
      "justifyContent": "center",
      "opacity": 0,
      "position": "absolute",
      "width": "100%",
    }
  }
  testID="overlayAnimatedView"
>
  <Text
    testID="text"
  >
    test
  </Text>
</View>
`;

exports[`Overlay Overlayが正常にrenderできることを確認: フェードアウト後 1`] = `null`;

exports[`Overlay Overlayが正常にrenderできることを確認: フェードイン後 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(0, 0, 0, 0.2)",
      "bottom": 0,
      "flex": 1,
      "height": 1334,
      "justifyContent": "center",
      "opacity": 1,
      "position": "absolute",
      "width": "100%",
    }
  }
  testID="overlayAnimatedView"
>
  <Text
    testID="text"
  >
    test
  </Text>
</View>
`;
